### DISCUSSION TITLE
Ranges

### DISCUSSION THREAD

**Discussion by ferdinandkeller** [time-rs/time#523] [Question] [answered]
I looked everywhere in the documentation, but it looks like there isn't a way to store ranges:
- You can use [`OffsetDateTime`](https://time-rs.github.io/api/time/struct.OffsetDateTime.html) to describe a specific date & time (10 of Jan 2022 at 13:06).
- You can use [`Duration`](https://time-rs.github.io/api/time/struct.Duration.html) to describe how long something takes (1 minute, 10 secondes, whatever).

But there is not struct to describe "Event E happens from date&time T1 to date&time T2", i.e. a range. And with this, there is no way to do simple things like:
- Is this date in the range ?
- What is the first monday of the range ?
- Split this range into intervals of one week.
- Split this range at the begining of every new week.
- ...

I require these functions for some project at work, so I'm probably not the only one needing these. So, is it not implemented because:
- These features are useful, but will never be implemented in the crate for reasons X and Y.
- These features could be in the crate, but nobody took time to work on it for now.

I am currently working on implementing these, because as I said, I **need** them. Should I make it a pull request ?

**Comment by jhpratt** [comment, marked as answer]
I have thought about this before, but not very much. I am open to the idea. As you're working on it anyway, feel free to open a pull request. Worst case is I would prefer a better API. Ultimately I'd like to use the unstable `Step` trait for some functionality, but there is no progress on stabilization there.

With regard to whether a value is in a certain range, you can actually use `(a..b).contains(&c)`.

**Reply by ferdinandkeller** [reply]
I settled down on using `Range<OffsetDateTime>` to represent ranges. This implementation offered many advantages, though it also has one little problem : We can't make sure when creating the range that both dates have the same offset. Should ranges of `OffsetDateTime` with different timezones be allowed ?

Here is an example, using a `days()` function on a `Range<OffsetDateTime>`:
```rust
let start = OffsetDateTime::now_utc();
let end = (start + Duration::days(20)).to_offset(offset!(+3));

for day in (start..end).days() {
    println!("{day}");
}
```

Should this be allowed, or should the `days()` function panic ? I think the day should have the same offset as the parent `start` and `end`, and in the case they are different, it should use either one.

If you have a different opinion on the matter, I'm open.

**Reply by jhpratt** [reply]
I think it's reasonable to have the offset always been either the start or end. My intuition leans towards the start, but if you have a reason to prefer the end, go for it. With that said, I think it should also be considered unspecified.
